### PR TITLE
Implement taubin smoothing

### LINKED ISSUE #2814: Add support for Taubin smoothing
### Describe the feature you would like to be added.

It would be great if pyvista supports non-shrinking Taubin smoothing. Reference: [here](https://dl.acm.org/doi/pdf/10.1145/218380.218473), or [here](https://doi.org/10.1109/ICCV.1995.466848)

In VTK, it is implemented in the class [vtkWindowedSincPolyDataFilter](https://vtk.org/doc/nightly/html/classvtkWindowedSincPolyDataFilter.html)

An undesirable effect of Laplace smoothing is the shrinking of the (closed) input surfaces. Taubin smoothing can avoid this effect by alternately applying shrinking and expanding smoothing. Find below some sample code.

The sample code makes use of this file: [geometry.vtp.zip](https://github.com/pyvista/pyvista/files/8931280/geometry.vtp.zip)

```python
import vtk
import pyvista as pv

mesh = pv.read("geometry.vtp")

# Use an excessive amount of smoothing to emphasize the effect.
# (It's a bit slow)
n_iter = 10000

# Laplace smoothing
mesh_lap = mesh.smooth(n_iter=n_iter, relaxation_factor=0.01)

# Taubin smoothing
smoothed = vtk.vtkWindowedSincPolyDataFilter()
smoothed.SetInputData(mesh)
smoothed.SetNumberOfIterations(n_iter)
smoothed.SetPassBand(0.1)
smoothed.SetNormalizeCoordinates(False)
smoothed.Update()

# Visualization
p = pv.Plotter(shape=(1,2))
p.subplot(0, 0)
p.add_title("Laplace")
p.add_mesh(mesh, opacity=0.4, color="white")
p.add_mesh(mesh_lap, color="red")

p.subplot(0, 1)
p.add_title("Taubin")
p.add_mesh(mesh, opacity=0.4, color="white")
p.add_mesh(smoothed.GetOutput(), color="blue")
p.link_views()
p.show()
```

Screenshot

### Links to VTK Documentation, Examples, or Class Definitions.

[vtkWindowedSincPolyDataFilter](https://vtk.org/doc/nightly

### Code of Conduct

- [x] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by akaszynski**
Resolves #2814 by implementing Taubin smoothing.

```py
import pyvista as pv
from pyvista import examples
mesh = examples.download_foot_bones().subdivide(2)
smoothed_mesh = mesh.smooth_taubin()
pl = pv.Plotter(shape=(1, 2))
_ = pl.add_mesh(mesh)
pl.add_text('Original Mesh')
pl.subplot(0, 1)
_ = pl.add_mesh(smoothed_mesh)
pl.add_text('Smoothed Mesh')
pl.show()
```

tmp

### CODE REVIEW

File: pyvista/core/filters/poly_data.py [hunk 1]
```diff
@@ -744,7 +747,126 @@ def smooth(
         if inplace:
             self.overwrite(mesh)
             return self
+        return mesh
+
+    def smooth_taubin(
+        self,
+        n_iter=20,
+        pass_band=0.1,
+        convergence=0.0,
```
Comment (akaszynski) [inline pyvista/core/filters/poly_data.py:756]:
```suggestion
```

File: pyvista/core/filters/poly_data.py [hunk 2]
```diff
@@ -744,7 +747,126 @@ def smooth(
...
+            boundary edges remain fixed as defined by ``feature_angle`` and
+            ``edge_angle``. Default ``False``.
+
+        normalize_coordinates : bool, optional
+            Flag to control coordinate normalization. To improve the
+            numerical stability of the solution and minimize the scaling the
+            translation effects, the algorithm can translate and scale the
+            position coordinates to within the unit cube ``[-1, 1]``, perform the
+            smoothing, and translate and scale the position coordinates back to
+            the original coordinate frame. This defaults to ``False``
```
Comment (akaszynski) [inline pyvista/core/filters/poly_data.py:811]:
```suggestion
            the original coordinate frame. This defaults to ``False``.
```

File: pyvista/core/filters/poly_data.py [hunk 3]
```diff
@@ -744,7 +747,125 @@ def smooth(
...
+        inplace=False,
+        progress_bar=False,
+    ):
+        """Smooth a PolyData DataSet with Taubin smoothing.
+
+        This filter allows you to smooth the mesh as in the Laplacian smoothing
+        implementation in :func:`smooth() <PolyDataFilters.smooth>`. However,
+        unlike Laplacian smoothing the surface does not "shrink" since this
+        filter relies on an alternative approach to smoothing. This filter is
+        more akin to low pass filter where undesirable high frequency features
```
Comment (tkoyama010) [inline pyvista/core/filters/poly_data.py:770]:
```suggestion
        more akin to a low pass filter where undesirable high frequency features
```

File: pyvista/core/filters/poly_data.py [hunk 4]
```diff
@@ -744,7 +747,125 @@ def smooth(
...
+
+        Parameters
+        ----------
+        n_iter : int, optional
+            This is the degree of the polynomial used to approximate the
+            windowed sync function. This is generally much less than the number
+            needed by :func:`smooth() <PolyDataFilters.smooth>`.
+
+        pass_band : float, optional
+            The passband value for the windowed sinc filter. This should be
```
Comment (tkoyama010) [inline pyvista/core/filters/poly_data.py:785]: I didn't know [sinc filter](https://en.wikipedia.org/wiki/Sinc_filter).

File: tests/test_filters.py [hunk 5]
```diff
@@ -998,13 +998,26 @@ def test_delaunay_3d():
...
+
+
+def test_smooth_taubin(uniform):
     data = examples.load_uniform()
     vol = data.threshold_percent(30)
-    surf = vol.extract_geometry(progress_bar=True)
-    smooth = surf.smooth()
+    surf = vol.extract_geometry()
+    smooth = surf.smooth_taubin()
     assert np.any(smooth.points)
```
Comment (tkoyama010) [inline tests/test_filters.py:1016]: Will the points increase if we use `smooth_taubin`? If so, we can add it to the test.

**Review by tkoyama010 — CHANGES_REQUESTED** [review]

**Review by tkoyama010 — APPROVED** [review]
LGTM

**Comment by tkoyama010** [comment]
It requires https://github.com/pyvista/pyvista/pull/2853 to merge.

**Comment by akaszynski** [comment]
> It requires #2853 to merge.

Going to go without that following our 24H rule.

**Pull request merged**